File: data_structure/DisjointSetUnion/java/DisjointSetUnion.java
import java.util.*;

public class DisjointSetUnion {
    public enum OperationType {
        UNION,
        CONNECTED,
        FIND
    }

    public static final class Operation {
        public final OperationType type;
        public final int first;
        public final int second;

        public Operation(OperationType type, int first, int second) {
            this.type = type;
            this.first = first;
            this.second = second;
        }
    }

    public static List<String> DisjointSetUnion(int n, List<Operation> operations) {
        int[] parent = new int[n];
        int[] componentSize = new int[n];
        for (int node = 0; node < n; node++) {
            parent[node] = node;
            componentSize[node] = 1;
        }

        List<String> queryResults = new ArrayList<>();
        for (Operation operation : operations) {
            if (operation.type == OperationType.UNION) {
                int rootA = findRoot(parent, operation.first);
                int rootB = findRoot(parent, operation.second);

                if (rootA == rootB) {
                    continue;
                }

                if (componentSize[rootA] < componentSize[rootB] ||
                    (componentSize[rootA] == componentSize[rootB] && rootA > rootB)) {
                    int temp = rootA;
                    rootA = rootB;
                    rootB = temp;
                }

                parent[rootB] = rootA;
                componentSize[rootA] += componentSize[rootB];
            } else if (operation.type == OperationType.CONNECTED) {
                queryResults.add(findRoot(parent, operation.first) == findRoot(parent, operation.second) ? "true" : "false");
            } else {
                queryResults.add(Integer.toString(findRoot(parent, operation.first)));
            }
        }

        return queryResults;
    }

    private static int findRoot(int[] parent, int node) {
        if (parent[node] == node) {
            return node;
        }

        parent[node] = findRoot(parent, parent[node]);
        return parent[node];
    }
}

File: data_structure/DisjointSetUnion/python/DisjointSetUnion.py
from __future__ import annotations

from enum import Enum
from typing import NamedTuple


class OperationType(Enum):
    UNION = "union"
    CONNECTED = "connected"
    FIND = "find"


class Operation(NamedTuple):
    type: OperationType
    first: int
    second: int


def DisjointSetUnion(n: int, operations: list[Operation]) -> list[str]:
    parent = list(range(n))
    component_size = [1] * n

    def find_root(node: int) -> int:
        if parent[node] == node:
            return node

        parent[node] = find_root(parent[node])
        return parent[node]

    query_results: list[str] = []
    for operation in operations:
        if operation.type is OperationType.UNION:
            root_a = find_root(operation.first)
            root_b = find_root(operation.second)

            if root_a == root_b:
                continue

            if component_size[root_a] < component_size[root_b] or (
                component_size[root_a] == component_size[root_b] and root_a > root_b
            ):
                root_a, root_b = root_b, root_a

            parent[root_b] = root_a
            component_size[root_a] += component_size[root_b]
        elif operation.type is OperationType.CONNECTED:
            query_results.append("true" if find_root(operation.first) == find_root(operation.second) else "false")
        else:
            query_results.append(str(find_root(operation.first)))

    return query_results
